elfo-dumper/src/file_registry.rs

```diff
@@ -17,30 +17,56 @@ pub(crate) struct FileRegistry {
 }
 
 impl FileRegistry {
-    pub(crate) async fn open(&self, path: &str, force: bool) -> Result<()> {
-        let mut file = {
-            let mut files = self.files.lock();
-
-            if !files.contains_key(path) {
-                files.insert(path.to_string(), FileHandle::default());
-            }
+    fn insert_handle(path: &str, files: &mut FxHashMap<String, FileHandle>) -> FileHandle {
+        let contained = files.contains_key(path);
+        if !contained {
+            files.insert(path.to_string(), FileHandle::default());
+        }
 
-            files.get(path).unwrap().clone()
-        };
+        files.get(path).unwrap().clone()
+    }
 
-        if file.open(path, force).await? {
+    async fn open_inner(path: &str, force: bool, fh: &mut FileHandle) -> Result<()> {
+        if fh.open(path, force).await? {
             debug!(%path, "file opened");
         }
 
         Ok(())
     }
 
-    pub(crate) async fn acquire(&self, path: &str) -> FileHandle {
-        self.files
-            .lock()
-            .get(path)
-            .expect("file must be open already")
-            .clone()
+    pub(crate) async fn open(&self, path: &str, force: bool) -> Result<()> {
+        let mut file = {
+            let mut files = self.files.lock();
+            Self::insert_handle(path, &mut files)
+        };
+
+        Self::open_inner(path, force, &mut file).await?;
+        Ok(())
+    }
+
+    /// Open file for writing dump in it, closing previous one if
+    /// paths differ.
+    pub(crate) async fn acquire_for_write(&self, old_path: &str, path: &str) -> Result<FileHandle> {
+        let mut fh = {
+            let mut files = self.files.lock();
+            if old_path != path {
```

Comment:
We can move the open logic to `FileHandle` and store `FileHandle` for a long time in every actor (instead of `path` and `path_swap`), which makes the code both clearer and less error-prone